add optional wave height input to subroutines in icepack that calculate wave heights

erinethomas opened the pull request:
This PR adds an **optional**  wave height input parameter to the wave fracture subroutine, that allows it to use the wave height from a coupled wave model as opposed to Icepack calculating its own wave height.
This change has been tested in fully coupled E3SM simulations with active wave-sea ice coupling.

erinethomas committed:
add optional wave height input to wavefracspec

erinethomas committed:
add wave height optional argument to wave_dep_growth

erinethomas committed:
bug fix

eclare108213 commented on columnphysics/icepack_fsd.F90:
```diff
@@ -829,6 +829,9 @@ subroutine wave_dep_growth (local_wave_spec, &
          dwavefreq       ! wave frequency bin widths (s^-1)
 
+      real(kind=dbl_kind), intent(in), optional :: &
+         wave_height
```
Add comment with definition and units.  Is this always significant wave height, or would it sometimes be some other wave height?

erinethomas replied:
it would always be significant wave hight.

eclare108213 commented on columnphysics/icepack_wavefracspec.F90:
```diff
@@ -214,6 +214,9 @@ subroutine icepack_step_wavefracture(wave_spec_type,   &
          d_afsd_wave     ! change in fsd due to waves
 
+      real (kind=dbl_kind), intent(in), optional :: &
+         wave_height    ! wave height 
```
Add units to comment (and make the comment the same here and in the icepack_fsd.F90 declaration for consistency)

eclare108213 commented on columnphysics/icepack_wavefracspec.F90:
```diff
@@ -183,7 +183,7 @@ subroutine icepack_step_wavefracture(wave_spec_type,   &
                   aice,          vice,            aicen,     &
                   wave_spectrum, wavefreq,        dwavefreq, &
-                  trcrn,         d_afsd_wave)
+                  trcrn,         d_afsd_wave, wave_height) 
```
align columns

eclare108213 reviewed (commented):
This looks fine to me.  Since `wave_height` is added as an optional argument in an Icepack interface routine, then I believe its status (present or not) will be inherited in other subroutine calls following this one, and if the driver does not include it, the code should be backward compatible.  But since optional arguments continue to befuddle me, I'd like @apcraig to confirm that the use of `optional` and `if (present())` is correct in this PR (thank you). 

When this change is pulled back to the Consortium repo, there will need to be a change to the Icepack and CICE drivers to recognize the new option using a namelist flag. @erin do you have a flag for it in E3SM, or is it somehow hard-coded as part of the LRW compset?

proteanplanet reviewed (approved):
Assuming changes recommended by @eclare108213 are enacted, I approve this PR.

erinethomas committed:
adding units to sig. wave height variable comments

erinethomas commented:
> This looks fine to me. Since `wave_height` is added as an optional argument in an Icepack interface routine, then I believe its status (present or not) will be inherited in other subroutine calls following this one, and if the driver does not include it, the code should be backward compatible. But since optional arguments continue to befuddle me, I'd like @apcraig to confirm that the use of `optional` and `if (present())` is correct in this PR (thank you).
> 
> When this change is pulled back to the Consortium repo, there will need to be a change to the Icepack and CICE drivers to recognize the new option using a namelist flag. @erin do you have a flag for it in E3SM, or is it somehow hard-coded as part of the LRW compset?

@eclare108213 - as an optional variable, if the wave height is not provided, the code runs as it currently does: by calculating the wave height. no namelist setting is needed to control it.  In E3SM, we do have a driver namelist flag to control wave-ice coupling based on the compset selected by the user. All wave - sea ice coupling variables are controlled by this one namelist setting, the variables do not have individual namelist settings. In other words, it is not possible to pass the wave spectra and NOT the wave height in the E3SM setup, if you have an wave model, all wave variables are calculated by the wave model and passed to MPAS-SI.  Hope this answers your question.

eclare108213 commented:
> > This looks fine to me. Since `wave_height` is added as an optional argument in an Icepack interface routine, then I believe its status (present or not) will be inherited in other subroutine calls following this one, and if the driver does not include it, the code should be backward compatible. But since optional arguments continue to befuddle me, I'd like @apcraig to confirm that the use of `optional` and `if (present())` is correct in this PR (thank you).
> > When this change is pulled back to the Consortium repo, there will need to be a change to the Icepack and CICE drivers to recognize the new option using a namelist flag. @erin do you have a flag for it in E3SM, or is it somehow hard-coded as part of the LRW compset?
> 
> @eclare108213 - as an optional variable, if the wave height is not provided, the code runs as it currently does: by calculating the wave height. no namelist setting is needed to control it. In E3SM, we do have a driver namelist flag to control wave-ice coupling based on the compset selected by the user. All wave - sea ice coupling variables are controlled by this one namelist setting, the variables do not have individual namelist settings. In other words, it is not possible to pass the wave spectra and NOT the wave height in the E3SM setup, if you have an wave model, all wave variables are calculated by the wave model and passed to MPAS-SI. Hope this answers your question.

As a PR into E3SM, this behavior is fine, but for other users there will need to be a namelist option added to control it, since the default at the moment is to pass one thing and not the other (spectra, height).  Having important behavior like this controlled only by the presence of an optional argument is a little risky, not good coding practice in my opinion.  My preference would be to go ahead and add the namelist (and associated documentation) to Icepack now.  I'm willing to do that and PR it into your branch.  Then you could choose whether to add it as a namelist in MPAS-SI or simply hardwire it when you initialize Icepack's parameters.  Would that be okay?

erinethomas commented:
@eclare108213 - yes, I will add a namelist option so the lines of code that check for the existence of a passed wave height field use both present(wave_height) AND a namelist flag. Code modification coming soon.

erinethomas committed:
adding wave_height_type namelist option to control wave_height_coupling.

eclare108213 commented on columnphysics/icepack_parameters.F90:
```diff
@@ -326,6 +326,9 @@ module icepack_parameters
       character (len=char_len), public :: &
          wave_spec_type = 'constant'  ! 'none', 'constant', or 'random'
+      
+      character (len=char_len), public :: &
```
This isn't super important, but wave_height_type and wave_height_spec could be listed under the same character declaration (here and elsewhere)

erinethomas replied:
done.

eclare108213 commented on columnphysics/icepack_wavefracspec.F90:
```diff
@@ -257,11 +261,18 @@ subroutine icepack_step_wavefracture(wave_spec_type,   &
+            local_sig_ht = wave_height
+         else 
+            write(warnstr,*) subname, &
+              'WARNING: Wave Height data not provided- calculating wave height internally'
```
Now that there's a namelist option, this case should abort instead of resetting the option to something else.

eclare108213 commented on configuration/scripts/icepack_in:
```diff
@@ -119,6 +119,7 @@
     saltflux_option = 'constant'
     oceanmixed_ice  = .true.
     wave_spec_type  = 'none'
+    wave_height_type= 'internal'
```
Since wave_spec_type = none, this should also be none.

eclare108213 commented on columnphysics/icepack_wavefracspec.F90:
```diff
@@ -257,11 +261,18 @@ subroutine icepack_step_wavefracture(wave_spec_type,   &
-         local_sig_ht = wave_height 
-      else
+        ! Add option to use wave height from wave model or file
+      if (trim(wave_height_type) == 'internal') then
```
There needs to be a wave_height_type = 'none' option, and if I'm reading the logic correctly, that could just be added here:
`if (... == 'internal' .or. ... == 'none') then`
(Is that right?)

eclare108213 commented on configuration/driver/icedrv_init.F90:
```diff
@@ -277,6 +280,7 @@ subroutine input_data
                                   ! 'mm_per_sec' = 'mks' = kg/m^2 s
       oceanmixed_ice  = .false.   ! if true, use internal ocean mixed layer
       wave_spec_type  = 'none'    ! type of wave spectrum forcing
+      wave_height_type= 'internal'! type of wave height forcing
```
set this default to 'none'

eclare108213 commented on configuration/driver/icedrv_step.F90:
```diff
@@ -668,17 +668,20 @@ subroutine step_dyn_wave (dt)
       character(len=*), parameter :: subname = '(step_dyn_wave)'
 
       call icepack_query_parameters(wave_spec_type_out=wave_spec_type)
+      call icepack_query_parameters(wave_height_type_out=wave_height_type)
```
both parameters can be included in the same query call

eclare108213 commented on doc/source/user_guide/ug_case_settings.rst:
```diff
@@ -390,6 +390,9 @@ forcing_nml
    "", "``random``", "wave data file is provided, sea surface height generated using random number (multiple iterations of wave fracture)", ""
+   "``wave_height_type``", "``internal``", "significant wave heights are calculated by icepack from the wave_spectra", "``internal``"
+   "", "``coupled``", "significant wave heights data provided from a coupled wave model, like WW3", ""
+   "", "``file``", "significant wave height data provided in an external file", ""
```
I'd leave the file option out for now, since we don't have one. The 'none' option comment could be something like "significant wave height not provided, instead calculated as internal"

eclare108213 reviewed (changes requested):
Thanks for updating the icepack driver too.  Did you run icepack tests in standalone mode?  If not, I can help ...

eclare108213 commented:
There is also some documentation that needs to be updated here: [doc/source/science_guide/sg_itd.rst](https://github.com/E3SM-Project/Icepack/blob/aff40ce8af3244ba698203dc5607204030fd6fb1/doc/source/science_guide/sg_itd.rst#L152)

eclare108213 commented:
Are you doing something to change this line?  
[configuration/driver/icedrv_step.F90](https://github.com/E3SM-Project/Icepack/blob/aff40ce8af3244ba698203dc5607204030fd6fb1/configuration/driver/icedrv_step.F90#L483)
Edit:  remove question re E3SM -- this IS the E3SM fork of Icepack.

erinethomas commented:
> Are you doing something to change this line? [configuration/driver/icedrv_step.F90](https://github.com/E3SM-Project/Icepack/blob/aff40ce8af3244ba698203dc5607204030fd6fb1/configuration/driver/icedrv_step.F90#L483) Edit: remove question re E3SM -- this IS the E3SM fork of Icepack.

I am not changing this line - since E3SM version of the code does NOT use the icepack driver (I have modified the driver used in the E3SM version)  however, people using the stand alone/icepack driver probably will need to modify this if they try to run icepack coupled to a wave model?  Happy to discuss the options needed here for stand alone runs.

erinethomas commented:
> There is also some documentation that needs to be updated here: [doc/source/science_guide/sg_itd.rst](https://github.com/E3SM-Project/Icepack/blob/aff40ce8af3244ba698203dc5607204030fd6fb1/doc/source/science_guide/sg_itd.rst#L152)

I do not see any issues with this documentation. The way significant wave height is calculated (internally or passed from a wave model) does not impact the way the FSD gets updated (from breaking). How exactly would you like me to update this documentation?

erinethomas committed:
clean up coupled wave_height code options

erinethomas committed:
fix abort when noe wave hiehgt data provided in fsd

erinethomas commented:
@eclare108213 - I have addressed most of your comments- a second round of review would be appreciated! (Also I may need help testing this in a stand alone icepack case)

erinethomas committed:
bug fix

erinethomas committed:
tidy up

erinethomas committed:
bug fix to abort

eclare108213 commented on columnphysics/icepack_parameters.F90:
```diff
@@ -323,9 +323,10 @@ module icepack_parameters
       logical (kind=log_kind), public :: &
          wave_spec = .false.          ! if true, use wave forcing
-
+      
```
remove new whitespace

eclare108213 commented on columnphysics/icepack_parameters.F90:
```diff
@@ -864,6 +865,9 @@ subroutine icepack_init_parameters(   &
          wave_spec_type_in  ! type of wave spectrum forcing
 
+      character (len=*), intent(in), optional :: &
+         wave_height_type_in  ! type of wave height forcing
```
this line can also be combined with the declaration above (again, not super important, just cosmetic)

eclare108213 commented on columnphysics/icepack_parameters.F90:
```diff
@@ -1884,6 +1889,9 @@ subroutine icepack_query_parameters(   &
          wave_spec_type_out ! type of wave spectrum forcing
 
+      character (len=*), intent(out), optional :: &
+         wave_height_type_out ! type of wave height forcing
```
combine with declaration above

eclare108213 commented on columnphysics/icepack_wavefracspec.F90:
```diff
@@ -254,7 +260,24 @@ subroutine icepack_step_wavefracture(wave_spec_type,   &
+            call icepack_warnings_add(subname//&
+            ' FSD needs wave_height_type=internal or coupled') 
+            call icepack_warnings_setabort(.true.,__FILE__,__LINE__)
+
```
remove blank line

eclare108213 commented on configuration/driver/icedrv_init.F90:
```diff
@@ -176,7 +177,8 @@ subroutine input_data
         atmiter_conv,    calc_dragio,     congel_freeze,   &
         tfrz_option,     saltflux_option, ice_ref_salinity, &
-        default_season,  wave_spec_type,  cpl_frazil,      &
+        default_season,  wave_spec_type, wave_height_type, & 
```
align column

eclare108213 commented on configuration/driver/icedrv_step.F90:
```diff
@@ -479,8 +483,9 @@ subroutine step_therm2 (dt)
-            if (tr_fsd) &
-            wave_sig_ht(i) = c4*SQRT(SUM(wave_spectrum(i,:)*dwavefreq(:)))
+            if (tr_fsd .and. trim(wave_height_type) == 'internal') then 
+                wave_sig_ht(i) = c4*SQRT(SUM(wave_spectrum(i,:)*dwavefreq(:)))
```
If I'm looking in the right place, it looks like you've simply commented this out in MPAS-seaice.  Is that right?  If so, maybe just add a comment after this line, like
`! else wave_sig_ht is provided by the coupler or other external data`

erinethomas replied:
Yes this is correct. MPAS-SI will only every get wave hight from WW3- and passes it straight into step_therm2... 
good suggestion.

eclare108213 commented on columnphysics/icepack_wavefracspec.F90:
```diff
@@ -254,7 +260,24 @@ subroutine icepack_step_wavefracture(wave_spec_type,   &
+         endif
+      elseif (trim(wave_height_type) == 'none') then
+            call icepack_warnings_add(subname//&
+            ' FSD needs wave_height_type=internal or coupled') 
```
Sorry, I remembered that we have FSD test cases for Icepack, and apparently `wave_spec_type='none'` does not abort.  The only place where `local_sig_ht` is used is in the immediately following conditional, so here I think simply setting `local_sig_ht = c0` will accomplish the same thing as in the original code. (Remove the warning and abort lines -- sorry for the wild goose chase!)

The two FSD test cases need to include the new namelist flag:

> ./configuration/scripts/options/set_nml.fsd1:wave_spec_type = 'none'

add `wave_height_type = 'none'`

> ./configuration/scripts/options/set_nml.fsd12:wave_spec_type = 'constant'

add `wave_height_type = 'internal'`
Hopefully that works.

erinethomas replied:
@eclare108213, Im pretty sure setting local_sig_ht = 0 will not do the same thing as the original code... the original code (with no wave height information) calculates the local_sig_ht from the provided spectrum ... so I'm not sure what to do with this new 'wave_height_type=none' option since the default icepack behavior (wave spectrum information is provided, but no wave heights) should be the same behavior as when we now set  'wave_height_type=internal'

erinethomas replied:
follow up comment: this subroutine never gets called in situations where 'wave_spec_type= none'.... 
so we need to determine what to do if a user sets "wave_spec_type /= none" and "wave_height_type = none"  (ideally settings of "none" are used in both options to indicate that NO wave forcing data provided- ie. both settings == 'none' or both settings != 'none') ...all this to say I'm fine with calling abort in the situation where wave_spec_type /= none but wave_height_type = none...

eclare108213 reviewed (commented):
All my suggestions here are cosmetic except how to handle the `wave_height_type='none'` case.  That needs to be done a little differently to maintain our current tests.  If you'll commit this then I'll set up a standalone Icepack test suite to double check things are working as expected.

erinethomas committed:
clean up

erinethomas committed:
new comment for clarity

erinethomas committed:
modify fsd tests for new wave_height_type

erinethomas committed:
fix namelist checking for wave_height_type

erinethomas commented:
@eclare108213 - ready for standalone testing. I am testing new icepack namelist settings in the coupled e3sm setup.

eclare108213 reviewed (changes requested):
`endif` needs to be added at line 666 in icedrv_init.F90.  With that change, base_suite tests all pass, including those with fsd turned on:

```
215 of 215 tests PASSED
0 of 215 tests PENDING
0 of 215 tests MISSING data
0 of 215 tests FAILED

testsuite.e3sm_waveht eclare$ ./results.csh | grep fsd
PASS conda_macos_smoke_col_1x1_debug_fsd12_run1year_short build
PASS conda_macos_smoke_col_1x1_debug_fsd12_run1year_short run
PASS conda_macos_smoke_col_1x1_debug_fsd12_run1year_short test
PASS conda_macos_smoke_col_1x1_debug_fsd12_run1year_short compare e3sm_base
PASS conda_macos_smoke_col_1x1_debug_fsd1_run1year build
PASS conda_macos_smoke_col_1x1_debug_fsd1_run1year run
PASS conda_macos_smoke_col_1x1_debug_fsd1_run1year test
PASS conda_macos_smoke_col_1x1_debug_fsd1_run1year compare e3sm_base
PASS conda_macos_restart_col_1x1_fsd12_short build
PASS conda_macos_restart_col_1x1_fsd12_short initialrun
PASS conda_macos_restart_col_1x1_fsd12_short run 
PASS conda_macos_restart_col_1x1_fsd12_short test 
PASS conda_macos_restart_col_1x1_fsd12_short compare e3sm_base
```

For the record, here's what I did to run these these tests:

create baseline:
```
git clone https://github.com/E3SM-Project/icepack icepack_e3sm_20251116
cd icepack_e3sm_20251116
./icepack.setup --suite base_suite -m conda -e macos -p 1x1 --testid e3sm_base --bgen e3sm_base 
```

clone wave code and test against baseline:
```
git clone https://github.com/erinethomas/icepack -b optional_WaveHeight icepack_e3sm_waveht
cd icepack_e3sm_waveht
./icepack.setup --suite base_suite -m conda -e macos -p 1x1 --testid e3sm_waveht --bcmp e3sm_base 
```

eclare108213 commented on configuration/driver/icedrv_init.F90:
```diff
@@ -646,10 +650,26 @@ subroutine input_data
+          (trim(wave_height_type)/='none'))) then
+         write (nu_diag,*) 'WARNING: Wave_spec_type=none, wave_height_type must also = none'
+         call icedrv_system_abort(file=__FILE__,line=__LINE__)
+      if (tr_fsd .and. &
```
add `endif` here

erinethomas replied:
good catch - thanks for doing the stand alone testing!

erinethomas committed:
add endif
